### DISCUSSION TITLE
"&mut self" function parameter in autogenerated trait

### DISCUSSION THREAD

**Discussion by jwinarske** [diwic/dbus-rs#309] [Q&A] [answered]
Hi.  Thanks for the great work.

Is there an option to indicate a trait function should have a "&mut self" parameter opposed to the default "&self"?

Such that one doesn't have to post edit the auto-generated file.

Thanks!

**Comment by diwic** [comment]
Hi,

Could you elaborate on which use case this would be helpful for you? What parameters are you running codegen with?

**Reply by jwinarske** [reply]
I'm using Default::default() for options.  It's generating the server interface, which I want.  I have it all wired together and am able to make calls, generate signals, and get responses via D-Feet.

Xml snippet:
```
        <method name="GetProfile">
            <arg name="profile" type="s" direction="out"/>
        </method>
        <method name="SetProfile">
            <arg name="profile" type="s" direction="in"/>
        </method>
        <signal name="ProfileSwitched">
            <arg name="profile" type="s"/>
        </signal>
```

With a generated interface looking like this:
```
pub trait MyConfig {
    fn get_profile(&self) -> Result<String, tree::MethodErr>;
    fn set_profile(&self, profile: &str) -> Result<(), tree::MethodErr>;
}
```
The following edit (+ others) is required to make the member variable mutable:
```
pub trait MyConfig {
    fn get_profile(&self) -> Result<String, tree::MethodErr>;
    fn set_profile(&mut self, profile: &str) -> Result<(), tree::MethodErr>;
}
```
Then implementing as
```
impl MyConfig for ProfileService {
    fn get_active_profile(&self) -> Result<String, MethodErr> {
        Ok(self.get_profile().clone())
    }
    fn set_active_profile(&mut self, profile: &str) -> Result<(), MethodErr> {
        Ok(self.set_profile(profile)) // writes self member variable
    }
```


Maybe there is a better way?

**Comment by diwic** [comment]
The problem here is that the way dbus-tree is designed, it does not have access to `&mut self`, only `&self` when it's supposed to call the method. 
This and other limitations have been resolved in dbus-crossroads, but there is no codegen for that (yet). I should probably write something, but it's fairly ergonomic to write server code without codegen in dbus-crossroads...

**Comment by jwinarske** [comment]
I see.  I'll take a look at dbus-crossroads.  Thanks!

**Comment by diwic** [comment, marked as answer]
Hi @jwinarske!

I've written an implementation of codegen for dbus-crossroads now, just pushed to git master. Would you like to try it out and see if it fits your needs? When using the crossroads option, the trait will take `&mut self` instead of `&self`.

**Reply by jwinarske** [reply]
Hi @diwic!

I like it.  An elegant solution!

It looks like `use dbus_crossroads as crossroads;` is missing from the autogenerated file.  Let me know when you push the change, and I will confirm it.

Thanks!

**Reply by diwic** [reply]
Are you sure the row is missing? It seems to work here:

    $ cd dbus-codegen
    $ cargo run -- --crossroads -d org.freedesktop.DBus | grep dbus_crossroads
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/dbus-codegen-rust --crossroads -d org.freedesktop.DBus`
    use dbus_crossroads as crossroads;

**Reply by jwinarske** [reply]
I just changed it be an "include", which resolves it for me.

Do you have an example of using crossroads with Tokio by any chance?

**Reply by diwic** [reply]
> I just changed it be an "include", which resolves it for me.

Sorry, could you explain what you mean by this?

 > Do you have an example of using crossroads with Tokio by any chance?

https://github.com/diwic/dbus-rs/blob/master/dbus-tokio/examples/tokio_adv_server_cr.rs and https://github.com/diwic/dbus-rs/blob/master/dbus-tokio/examples/tokio_server_cr.rs
Neither uses codegen (yet).

**Reply by jwinarske** [reply]
`include!("config_service/com_example_config.rs");`

vs
```pub use com_example_config::{register_com_example_config, ComExampleConfig};

mod com_example_config;
```

This is what I have so far (not tested): [simple-service](https://github.com/jwinarske/simple-service-rs)

**Reply by diwic** [reply]
@jwinarske  Here's what I did to make your code compile: https://gist.github.com/diwic/ea2760b24a6246cb707e880188553363

**Reply by jwinarske** [reply]
Thanks David!
